opencog/atoms/distvalue/DistributionalValue.cc

```diff
@@ -0,0 +1,455 @@
+
+#include <opencog/util/numeric.h>
+
+#include <opencog/atoms/distvalue/DistributionalValue.h>
+#include <opencog/atomspace/AtomSpace.h>
+
+#include <boost/range/combine.hpp>
+#include <iomanip>
+
+using namespace opencog;
+
+count_t DistributionalValue::DEFAULT_K = 800.0;
+
+DistributionalValue::DistributionalValue()
+	: Value(DISTRIBUTIONAL_VALUE)
+{}
+
+DistributionalValue::DistributionalValue(const DVCounter &dvctr)
+	: Value(DISTRIBUTIONAL_VALUE)
+{
+	for (auto elem : dvctr)
+	{
+		if (elem.second <= 0)
+			throw std::invalid_argument("DV cant have an element with count 0.");
+	}
+	_value = dvctr;
+}
+
+//Create a DV from the Parameters of a SimpleTV
+//Not recommended as it results in a DV with only 1 singleton set
+//which in some calculations is unusalbe as the chance of overlaps in the
+//Sets/Intervals is small
+DistributionalValue::DistributionalValue(double mode,double conf)
+	: Value(DISTRIBUTIONAL_VALUE)
+{
+	confidence_t cf = std::min(conf, 0.9999998);
+	double count = (DEFAULT_K * cf / (1.0 - cf));
+	//DV Interval with count 0 is undefined
+	count = std::max(count, 0.0000002);
+	DVKey k{Interval{mode}};
+	_value[k] = count;
+}
+
+DistributionalValuePtr DistributionalValue::createDV(const DVCounter &dvctr)
+{
+	return std::make_shared<const DistributionalValue>(dvctr);
+}
+
+DistributionalValuePtr DistributionalValue::createDV(double mode,
+                                                     double conf)
+{
+	return std::make_shared<const DistributionalValue>(mode,conf);
+}
+
+DistributionalValuePtr
+DistributionalValue::UniformDistributionalValue(const DVKey &k,int c)
+{
+	DVCounter dvctr;
+	dvctr[k] = c;
+	return createDV(dvctr);
+}
+
+
+DistributionalValuePtr
+DistributionalValue::UniformDistributionalValue(const DVKeySeq &ks,int c)
+{
+	DVCounter dvctr;
+	for (auto k : ks)
+	{
+		dvctr[k] = c;
+	}
+	return createDV(dvctr);
+}
+
+DistributionalValuePtr DistributionalValue::TRUE_TV()
+{
+	DVKey v1{Interval{1.0}};
+	DVKey v2{Interval{0.0}};
+	DVCounter dvc;
+	dvc[v1] = 1;
+	dvc[v2] = 0;
+	return std::make_shared<const DistributionalValue>(dvc);
+}
+DistributionalValuePtr DistributionalValue::FALSE_TV()
+{
+	DVKey v1{Interval{1.0}};
+	DVKey v2{Interval{0.0}};
+	DVCounter dvc;
+	dvc[v1] = 0;
+	dvc[v2] = 1;
+	return std::make_shared<const DistributionalValue>(dvc);
+}
+DistributionalValuePtr DistributionalValue::DEFAULT_TV()
+{
+	DVKey v1{Interval{1.0}};
+	DVKey v2{Interval{0.0}};
+	DVCounter dvc;
+	dvc[v1] = 0;
+	dvc[v2] = 0;
+	return std::make_shared<const DistributionalValue>(dvc);
+}
+
+bool DistributionalValue::is_uniform() const
+{
+	double val = _value.begin()->second;
+	for (auto p : _value)
+	{
+		if (val != p.second)
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+//Add Evidence for the provided key i.e increment the count of this Interval
+//Returns a new DVPtr with the update count
+DistributionalValuePtr DistributionalValue::add_evidence(const DVKey &h) const
+{
+	DVCounter newdvc = _value;
+	newdvc[h] += 1;
```

Comment:
why just plus-one?  It could be a positive or a negative double, in general.

Also, this cannot possibly work as written. Recall that `DVKey` is not a single bin, but is a list of bins.  So this doesn't make sense, as written.

There are also other issues:  after you fix `DVKey` to be `Bin` you still have the issue where `DVCounter` might be holding a bin [0.4, 0.5] bu you accidentally supplied a bin of width [0.40001, 0.500001] and so now you increment the count in a bin that completely overlaps an existing bin.  So this API is borken.